remove iot tlv from ccn-lite

mfrey opened the pull request:
This PR removes support for the IOT TLV format (and addresses #212). After the review and approval the commits **need to be squashed** before merging. I've basically tried to split the whole removal process into small pieces, so reverting (if necessary) should be easier.

cgundogan commented on src/ccnl-fwd/src/ccnl-echo.c:
```diff
@@ -71,8 +71,6 @@ ccnl_echo_request(struct ccnl_relay_s *relay, struct ccnl_face_s *inface,
     ucp = reply->data;
     len = reply->datalen;
 
-    while (!ccnl_switch_dehead(&ucp, &len, &enc)); // for iot2014 encoding
```
@blacksheeep I am not very familiar with this program. Is it safe to remove this line here (despite the comment saying it's just for iottlv, maybe the comment is wrong?)

blacksheeep replied:
Neither me, and we have a lack of documentation.

I think this functionality is only required to distinguish CISTLV and IOTTLV, so that it can be removed.

mfrey commented:
Hi, 

thanks for the review! Comments below


https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/ccnl-core/src/ccnl-frag.c#L49 (note about depcretation for IOT TLV?)

> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/doc/internal/README-multiprotocol.md

See comment

> https://github.com/mfrey/ccn-lite/blame/c28511cdec2bab17552cf165ad16a792c53ac3fb/README.md#L175
> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/CMakeLists.txt#L61
> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/build-test.mk#L81
> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/build-test.mk#L96
> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/ccn-lite-simu.c#L42
> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/ccnl-android/native/CMakeLists.txt#L44
> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/ccnl-android/native/src/ccn-lite-android.c#L685

Removed.

> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/ccnl-android/native/src/ccn-lite-jni.c#L153 (seems to be a bug here)

I would agree.

> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/ccnl-core/include/ccnl-defs.h#L98
> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/ccnl-core/include/ccnl-defs.h#L121
> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/ccnl-core/src/ccnl-interest.c#L90
> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/ccnl-core/src/ccnl-pkt.c#L61
> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/ccnl-core/src/ccnl-pkt.c#L112
> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/ccnl-core/src/ccnl-relay.c#L652
> https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/ccnl-lnxkernel/ccn-lite-lnxkernel.c#L42
>  https://github.com/mfrey/ccn-lite/blob/c28511cdec2bab17552cf165ad16a792c53ac3fb/src/doc/Doxyfile.in#L2064

Removed

cgundogan commented:
>The following format numbers have been assigned so far:
>
>  0  CCNL_ENC_CCNB
  1  CCNL_ENC_NDN2013
  2  CCNL_ENC_CCNX2014
  3  CCNL_ENC_IOT2014
  4  CCNL_ENC_LOCALRPC

These (still) match to
```c
enum {
  CCNL_ENC_CCNB,
  CCNL_ENC_NDN2013,
  CCNL_ENC_CCNX2014,
  CCNL_ENC_IOT2014,
  CCNL_ENC_LOCALRPC,
  CCNL_ENC_CISCO2015
};
```

In theory it should be fine just to remove `CCNL_ENC_IOT2014` and from the README. I assume that nobody used the actual number, but rather the enum constant to refer to the appropriate value.

What we can do is:
1. assign numbers to the CCNL_ENC_ enum explicitly
```c
enum {
  CCNL_ENC_CCNB = 0,
  CCNL_ENC_NDN2013 = 1,
  CCNL_ENC_CCNX2014 = 2,
  CCNL_ENC_LOCALRPC = 3,
  CCNL_ENC_CISCO2015 = 4
};
```
2. refer to the enum definition [in the code](https://github.com/cn-uofbasel/ccn-lite/blob/master/src/ccnl-core/include/ccnl-defs.h#L117). This would be so much simpler if we had a) doxygen for that enum and b) a compiled doc somewhere reachable from the webz. Just refering to the line number like I proposed before will break eventually if the file is touched. But we can go with that solution for now and adapt the README with a permanent link to the doxygen of that enum later.

blacksheeep commented:
@cgundogan 
We removed the numbers to be able to add new packet formats without confusion.
There shouldn't be trouble with that.
But yeah, doc would be good

mfrey commented:
I've added doxygen documentation to the enum and also added a typedef. I removed the enum members from the readme (unfortunately, I can't point to an online documentation) and made a remark where to find the enum. 

@cgundogan care to check the PR again?

cgundogan commented on doc/internal/README-multiprotocol.md:
```diff
@@ -15,7 +15,7 @@ potentially confronted with packets encoded in different schemas like
 among these packets because luckily enough the "type values" were
 sparsely populated and the programmers had chosen code points that do
-not overlap. But with the compact IOT2014 format, there is no sparsity
+not overlap. But with no packet formats, there might be no sparsity
```
>But with no packet formats, there might be no sparsity

Hmm, this sounds weird now ..

Maybe just remove this block:
>In the past it was possible to discriminate among these packets because luckily enough the "type values" were sparsely populated and the programmers had chosen code points that do not overlap. But with no packet formats, there might be no sparsity anymore. 

?

mfrey replied:
That's right. It was supposed to be "new" and not "no".

mfrey commented:
okay, it builds (again?). could someone please approve?
